Lib/compression/zstd/__init__.py

```diff
@@ -0,0 +1,293 @@
+"""Python bindings to Zstandard (zstd) compression library, the API style is
+similar to Python's bz2/lzma/zlib modules.
+"""
+
+__all__ = (
+    # From this file
+    "compressionLevel_values",
+    "get_frame_info",
+    "CParameter",
+    "DParameter",
+    "Strategy",
+    "finalize_dict",
+    "train_dict",
+    "zstd_support_multithread",
+    "compress",
+    "decompress",
+    # From _zstd
+    "ZstdCompressor",
+    "ZstdDecompressor",
+    "ZstdDict",
+    "ZstdError",
+    "get_frame_size",
+    "zstd_version",
+    "zstd_version_info",
+    # From zstd.zstdfile
+    "open",
+    "ZstdFile",
+)
+
+import enum
+import functools
+import dataclasses
+
+from compression.zstd.zstdfile import ZstdFile, open
+from _zstd import *
+
+import _zstd
+
+
+_ZSTD_CStreamSizes = _zstd._ZSTD_CStreamSizes
+_ZSTD_DStreamSizes = _zstd._ZSTD_DStreamSizes
+_train_dict = _zstd._train_dict
+_finalize_dict = _zstd._finalize_dict
+
+
+@dataclasses.dataclass(frozen=True)
+class _CompressionLevelValues:
+    default: int
+    min: int
+    max: int
+
+compressionLevel_values = _CompressionLevelValues(*_zstd._compressionLevel_values)
+
+@dataclasses.dataclass(frozen=True)
+class FrameInfo:
+    """A dataclass storing information about a Zstandard frame."""
+    decompressed_size: int
+    dictionary_id: int
+
+
+def get_frame_info(frame_buffer):
+    """Get zstd frame information from a frame header.
+
+    Parameter
+    frame_buffer: A bytes-like object. It should starts from the beginning of
+                  a frame, and needs to include at least the frame header (6 to
+                  18 bytes).
+
+    Return a FrameInfo dataclass, which currently has two attributes
+
+    'decompressed_size' is the size in bytes of the data in the frame when
+    decompressed.
+
+    If decompressed_size is None, decompressed size is unknown.
+
+    'dictionary_id' is a 32-bit unsigned integer value. 0 means dictionary ID was
+    not recorded in the frame header, the frame may or may not need a dictionary
+    to be decoded, and the ID of such a dictionary is not specified.
+    """
+
+    ret_tuple = _zstd._get_frame_info(frame_buffer)
+    return FrameInfo(*ret_tuple)
+
+
+def _nbytes(dat):
+    if isinstance(dat, (bytes, bytearray)):
+        return len(dat)
+    with memoryview(dat) as mv:
+        return mv.nbytes
+
+
+def train_dict(samples, dict_size):
+    """Train a zstd dictionary, return a ZstdDict object.
+
+    Parameters
+    samples:   An iterable of samples, a sample is a bytes-like object
+               represents a file.
+    dict_size: The dictionary's maximum size, in bytes.
+    """
+    # Check argument's type
+    if not isinstance(dict_size, int):
+        raise TypeError('dict_size argument should be an int object.')
```

Comment:
Are int subclasses permitted?
```suggestion
    if not isinstance(dict_size, int):
        ds_cls = type(dict_size).__qualname__
        raise TypeError('dict_size must be an int object, not {ds_cls!r}.')
```

Reply:
I don't see why not. Anything that can be passed to `PyLong_AsSsize_t` should work.